src/include/migraphx/op/group_query_attention.hpp

```diff
@@ -0,0 +1,588 @@
+
+        auto output_shape_0 = output_shape.sub_shapes().front();
+        argument result{output_shape_0};
+        argument qkv_rotary{
+            shape{output_shape_0.type(),
+                  {batch_size, num_heads + 2 * kv_num_heads, sequence_length, head_size}}};
+
+        shape kv_shape{output_shape_0.type(),
+                       {batch_size, kv_num_heads, past_sequence_length, head_size}};
+        argument present_k_out{kv_shape};
+        argument present_v_out{kv_shape};
+        argument attention_probs{shape{
+            output_shape_0.type(), {batch_size, num_heads, sequence_length, present_kv_seqlen}}};
+
+        args[0] = args[0].reshape(
+            shape{output_shape_0.type(),
+                  {batch_size, sequence_length, num_heads + 2 * kv_num_heads, head_size}});
+        argument qkv{qkv_rotary.get_shape()};
+        visit_all(qkv, args[0])([&](auto a, auto b) {
+            auto in_shape  = args[0].get_shape();
+            auto out_shape = qkv.get_shape();
+            shape_for_each(in_shape, [&](const auto& idx) {
+                std::vector<std::size_t> out_idx{idx[0], idx[2], idx[1], idx[3]};
+                a(out_idx.begin(), out_idx.end()) = b(idx.begin(), idx.end());
+            });
+        });
+
+        visit_all(result,
+                  qkv,
+                  args[3],
+                  args[4],
+                  args[7],
+                  args[8],
+                  qkv_rotary,
+                  present_k_out,
+                  present_v_out,
+                  attention_probs)([&](auto output,
+                                       auto query,
+                                       auto past_key,
+                                       auto past_value,
+                                       auto cos_cache,
+                                       auto sin_cache,
+                                       auto rotary_qkv,
+                                       auto present_k,
+                                       auto present_v,
+                                       auto attn_probs) {
+            visit_all(args[5])([&](auto seqlens_k) {
+                par_for(kv_shape.elements(), [&](auto i) {
+                    present_k[i] = past_key[i];
+                    present_v[i] = past_value[i];
+                });
+                auto seq_stride          = head_size;
+                auto head_stride         = sequence_length * seq_stride;
+                auto batch_stride        = num_heads + 2 * kv_num_heads;
+                auto position_ids_format = sequence_length == 1 ? 1 : 0;
+                bool transposed          = true;
+                std::vector<int64_t> pos_ids(sequence_length == 1 ? batch_size : 1);
+                if(sequence_length == 1)
+                {
+                    for(std::size_t b = 0; b < batch_size; b++)
+                    {
+                        pos_ids[b] = static_cast<int64_t>(seqlens_k[b]);
+                    }
```

Comment:
Can be written as `std::copy(seqlens_k.begin(), seqlens_k.begin() + batch_size, pos_ids.begin())`.